src/FSharp.Core/map.fs

```diff
@@ -882,6 +882,41 @@ type Map<[<EqualityConditionalOn>] 'Key, [<EqualityConditionalOn; ComparisonCond
     override this.GetHashCode() =
         this.ComputeHashCode()
 
+    interface IStructuralEquatable with
+        member this.Equals(that, comparer) =
+            match that with
+            | :? Map<'Key, 'Value> as that ->
+                use e1 = (this :> seq<_>).GetEnumerator()
+                use e2 = (that :> seq<_>).GetEnumerator()
+
+                let rec loop () =
```

Comment:
This isn't tail recursive, won't it stack overflow when there are too many elements? Does it get translated to a while loop?

Reply:
Why is not tail recursive? 
&& does not behave like e.g. addition in that regard.

Reply:
At first glance it looked like the recursive call wasn't the last operation there. But I guess with && short-circuiting it is?